File: app/src/main/java/com/saizad/mvvmexample/di/main/MainModule.java
package com.saizad.mvvmexample.di.main;

import android.content.SharedPreferences;

import com.google.gson.Gson;
import com.saizad.mvvm.ActivityResult;
import com.saizad.mvvm.CurrentUserType;
import com.saizad.mvvm.FCMToken;
import com.saizad.mvvm.NotifyOnce;
import com.saizad.mvvm.ObjectPreference;
import com.saizad.mvvm.SaizadEasyRetrofit;
import com.saizad.mvvmexample.api.MainApi;

import javax.inject.Named;
import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.components.SingletonComponent;
import io.reactivex.subjects.BehaviorSubject;
import sa.zad.easypermission.PermissionManager;

@Module
@InstallIn(SingletonComponent.class)
public class MainModule {

    @Singleton
    @Provides
    static MainEnvironment providesAuthEnvironment(FCMToken fcmToken, MainApi mainApi, PermissionManager permissionManager, CurrentUserType currentUser, BehaviorSubject<ActivityResult<?>> navigationFragmentResult, @Named("notification") BehaviorSubject<NotifyOnce<?>> notifyOnceBehaviorSubject) {
        return new MainEnvironment(fcmToken, mainApi, navigationFragmentResult, currentUser, permissionManager, notifyOnceBehaviorSubject);
    }

    @Singleton
    @Provides
    static MainApi providesMainApi(SaizadEasyRetrofit saizadEasyRetrofit) {
        return saizadEasyRetrofit.provideRetrofit().create(MainApi.class);
    }

    @Singleton
    @Provides
    static ObjectPreference providesAddPharmacyDraft(SharedPreferences sharedPreferences, Gson gson) {
        return new ObjectPreference(sharedPreferences, gson);
    }

}

File: app/src/main/java/com/saizad/mvvmexample/di/main/MainModule.kt
package com.saizad.mvvmexample.di.main

import android.content.SharedPreferences
import com.google.gson.Gson
import com.saizad.mvvm.*
import com.saizad.mvvmexample.api.MainApi
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import io.reactivex.subjects.BehaviorSubject
import sa.zad.easypermission.PermissionManager
import javax.inject.Named
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
object MainModule {
    
    @Singleton
    @Provides
    fun providesAuthEnvironment(
        mainApi: MainApi,
        fcmToken: FCMToken,
        permissionManager: PermissionManager,
        currentUser: CurrentUserType<*>,
        navigationFragmentResult: BehaviorSubject<ActivityResult<*>>,
        @Named("notification") notifyOnceBehaviorSubject: BehaviorSubject<NotifyOnce<*>>
    ): MainEnvironment {
        return MainEnvironment(
            mainApi,
            fcmToken,
            navigationFragmentResult,
            currentUser,
            permissionManager,
            notifyOnceBehaviorSubject
        )
    }
    
    @Singleton
    @Provides
    fun providesMainApi(saizadEasyRetrofit: SaizadEasyRetrofit): MainApi {
        return saizadEasyRetrofit.provideRetrofit().create(MainApi::class.java)
    }
    
    @Singleton
    @Provides
    fun providesObjectPreference(
        sharedPreferences: SharedPreferences, gson: Gson
    ): ObjectPreference {
        return ObjectPreference(sharedPreferences, gson)
    }
}